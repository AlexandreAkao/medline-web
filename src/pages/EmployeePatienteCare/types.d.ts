type AppointmentType = 'Em espera Triagem' | 'Triagem' | 'Esperando Atendimento' | 'Em Atendimento' | 'Atendido';
