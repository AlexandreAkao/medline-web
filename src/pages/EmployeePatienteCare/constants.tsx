import { AiOutlineEdit } from 'react-icons/ai';

export const COLUMNS = [
  {
    Header: 'Lugar na fila',
    accessor: 'firstName',
  },
  {
    Header: 'Nome',
    accessor: 'lastName',
  },
  {
    Header: 'PrevisÃ£o de atendimento',
    accessor: 'age',
  },
  {
    Header: 'Status',
    accessor: 'status',
  },
  {
    Header: '',
    accessor: 'action',
    Cell: () => <AiOutlineEdit />,
  },
];
