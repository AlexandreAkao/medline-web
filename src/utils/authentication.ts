import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
