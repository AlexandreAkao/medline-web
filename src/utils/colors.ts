export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
