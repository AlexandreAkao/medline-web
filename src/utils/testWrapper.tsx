import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component: React.ReactNode) =>
  render(<MemoryRouter>{component}</MemoryRouter>);
