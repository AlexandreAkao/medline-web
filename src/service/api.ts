import axios, { AxiosError } from 'axios';
import { toast } from 'react-toastify';

import { getToken } from 'utils/authentication';

const api = axios.create({
  baseURL: (import.meta.env.VITE_BFF_BASE_URL as string) ?? '',
});

api.interceptors.request.use(
  config => {
    const authorization = getToken() ? { Authorization: `Bearer ${getToken()}` } : null;
    return {
      ...config,
      headers: {
        ...authorization,
      },
    };
  },
  error => {
    return Promise.reject(error);
  },
);

api.interceptors.response.use(
  response => {
    return response;
  },
  (error: AxiosError) => {
    if (error.response?.status === 401) {
      axios
        .post<IRefreshToken>(
          'https://securetoken.googleapis.com/v1/token',
          {
            grant_type: 'refresh_token',
            refresh_token: localStorage.getItem('refresh_token') ?? '',
          },
          {
            params: {
              key: 'AIzaSyDYolHTkT0zuI3X1MNup36eSi2gZ3INa2g',
            },
          },
        )
        .then(data => {
          const { access_token: accessToken } = data.data;

          localStorage.setItem('token', accessToken);
        });
    } else if (error.response?.status !== 404)
      if (error.response && error.response.status >= 500)
        toast.error(error.message, { autoClose: 3000, theme: 'colored' });
    return Promise.reject(error);
  },
);

export default api;
