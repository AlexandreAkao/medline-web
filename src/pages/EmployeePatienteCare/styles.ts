import styled from 'styled-components';

import { TableContainer } from 'components/Table/styles';
import colors from 'styles/colors';

export const EmployeePatienteCareContainer = styled.div`
  background-color: ${colors.white.lighter};
  height: 100%;
  background-repeat: no-repeat;
  background-attachment: fixed;
  background-position-x: 50%;
  overflow: auto;
`;

export const EmployeePatienteCareMain = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 2rem auto;
  width: 80%;

  ${TableContainer} {
    width: 100%;

    svg {
      cursor: pointer;
    }
  }
`;

export const EmployeePatienteCareTitle = styled.h1`
  font-size: 2.125rem;
  font-weight: 500;
  margin-bottom: 2rem;
`;

export const EmployeePatienteCareModalContainer = styled.div`
  padding: 1.5rem;
  width: 70vw;

  > h2 {
    margin-bottom: 2rem;
  }
`;
