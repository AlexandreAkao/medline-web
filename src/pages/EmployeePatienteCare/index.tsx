import { useEffect, useState } from 'react';
import { toast } from 'react-toastify';

import MedlineHeader from 'components/Header/MedlineHeader';
import Table from 'components/Table';
import {
  EmployeePatienteCareContainer,
  EmployeePatienteCareMain,
  EmployeePatienteCareModalContainer,
  EmployeePatienteCareTitle,
} from 'pages/EmployeePatienteCare/styles';
import { COLUMNS, PATIENTE_STATUS } from 'pages/EmployeePatienteCare/constants';
import api from 'service/api';
import Modal from 'components/Modal';
import Dropdown from 'components/Dropdown';
import { useLoader } from 'hooks/useLoader';

function EmployeePatienteCare() {
  const [appointments, setAppointments] = useState<IAppointment[]>([]);
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [patienteSelected, setPatienteSelected] = useState<IAppointment>();
  const { setIsLoading } = useLoader();

  const handleOpenModal = (appointment: IAppointment) => {
    setIsVisible(true);
    setPatienteSelected(appointment);
  };

  const handleCloseModal = () => {
    setIsVisible(false);
    setPatienteSelected(undefined);
  };

  const updatePatientes = async () => {
    const patients = await api.get<IAppointment[]>('appointment');

    setAppointments(patients.data.filter(patient => patient.queue_position !== null));
  };

  const handleChangePatienteStatus = async (_event?: React.MouseEvent<HTMLButtonElement>, option?: IOption) => {
    try {
      setIsLoading(true);
      const status = PATIENTE_STATUS.findIndex(value => value.value === option?.value);
      const response = await api.patch(`appointment/${patienteSelected?.id}`, { status });

      if (response.status !== 200) {
        toast.error('Ocorreu um error');
        return;
      }

      updatePatientes();
      toast.success('Paciente atualizado com sucesso!');
      handleCloseModal();
    } catch (e) {
      toast.error('Ocorreu um error');
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    updatePatientes();
  }, []);

  return (
    <EmployeePatienteCareContainer>
      <MedlineHeader />
      <Modal isVisible={isVisible} onClose={handleCloseModal}>
        <EmployeePatienteCareModalContainer>
          <h2>Edite o status do paciente: {patienteSelected?.name}</h2>
          <Dropdown
            placeholder="Esolha um status"
            options={PATIENTE_STATUS.filter(value => value.value !== patienteSelected?.status)}
            buttonConfig={{ children: 'Atualizar', onClick: handleChangePatienteStatus }}
          />
        </EmployeePatienteCareModalContainer>
      </Modal>
      <EmployeePatienteCareMain>
        <EmployeePatienteCareTitle>Atendimento</EmployeePatienteCareTitle>
        <Table columns={COLUMNS(handleOpenModal)} data={appointments as unknown as IData[]} />
      </EmployeePatienteCareMain>
    </EmployeePatienteCareContainer>
  );
}

export default EmployeePatienteCare;
