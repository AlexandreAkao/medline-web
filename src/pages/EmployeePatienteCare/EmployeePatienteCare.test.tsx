import { screen } from '@testing-library/react';
import { vi } from 'vitest';

import { renderWithProvider } from 'utils/testWrapper';
import EmployeePatienteCare from 'pages/EmployeePatienteCare';
import api from 'service/api';

describe('EmployeePatienteCare', () => {
  beforeEach(() => {
    vi.spyOn(api, 'get').mockImplementation(() =>
      Promise.resolve({
        data: [],
      }),
    );
  });

  const renderComponent = () => {
    return renderWithProvider(<EmployeePatienteCare />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    renderComponent();
    const titleElement = screen.getByRole('heading');

    expect(titleElement).toHaveTextContent('Atendimento');
  });
});
