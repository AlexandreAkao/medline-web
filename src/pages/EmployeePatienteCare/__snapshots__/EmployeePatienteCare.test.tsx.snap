// Vitest Snapshot v1

exports[`EmployeePatienteCare > should match with snapshot 1`] = `
<div>
  <div
    class="sc-evZas cdVbbs"
  >
    <header
      class="sc-gsnTZi jLFYRz"
      data-testid="test-navigation"
    >
      <img
        alt="Medline logo"
        class="sc-dkzDqf dxiqfB"
        src="/src/assets/images/medline-logo.svg"
      />
      <nav
        class="sc-hKMtZM iKwNjC"
        data-testid="test-navigation-items"
      >
        <div
          class="sc-bczRLJ erPxbe"
        >
          <button
            class="sc-jSMfEi fWVToL"
            type="button"
          >
            Cadastro
          </button>
        </div>
        <div
          class="sc-bczRLJ erPxbe"
        >
          <a
            href="/login"
          >
            Entrar
          </a>
        </div>
      </nav>
    </header>
    <div
      class="sc-breuTD leQAZN"
    >
      <h1
        class="sc-ksZaOG fvoegn"
      >
        Atendimento
      </h1>
      <div
        class="sc-crXcEl fGbdKS"
      >
        <table
          role="table"
        >
          <thead>
            <tr
              role="row"
            >
              <th
                colspan="1"
                role="columnheader"
              >
                Lugar na fila
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                Nome
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                Previs√£o de atendimento
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                Status
              </th>
              <th
                colspan="1"
                role="columnheader"
              />
            </tr>
          </thead>
          <tbody
            role="rowgroup"
          />
        </table>
      </div>
    </div>
  </div>
  ,
</div>
`;
